### PR TITLE
Adding Proximal Gradient Descent

### PR DESCRIPTION
**Pull request by kavyasrinet**
Adding proximal gradient descent and test:
prox_param = param - learning_rate * grad
param = sign(prox_param) / (1 + learning_rate * l2) *
        max { |prox_param| - learning_rate * l1 , 0 }

### CODE REVIEW

File: paddle/operators/proximal_gd_op.cc [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+};
+
+class ProximalGDOpMaker : public framework::OpProtoAndCheckerMaker {
+ public:
+  ProximalGDOpMaker(framework::OpProto *proto,
+                    framework::OpAttrChecker *op_checker)
+      : OpProtoAndCheckerMaker(proto, op_checker) {
+    AddInput("Param",
+             "(Tensor, default Tensor<float>) "
+             "Input parameter value that has to be updated");
```
Comment (lcy-seso) [inline paddle/operators/proximal_gd_op.cc]: I think it is better to add periods at the end of the comment. Below are the same.

File: paddle/operators/proximal_gd_op.cu [hunk 2]
```diff
@@ -0,0 +1,20 @@
+/* Copyright (c) 2016 PaddlePaddle Authors. All Rights Reserve.
```
Comment (lcy-seso) [inline paddle/operators/proximal_gd_op.cu]: The indentation of the license has the problem. The correct indentation is like that in `proximal_gd_op.cc`.

File: paddle/operators/proximal_gd_op.h [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+template <typename Place, typename T>
+class ProximalGDOpKernel : public framework::OpKernel<T> {
+ public:
+  void Compute(const framework::ExecutionContext& ctx) const override {
+    auto* param_out = ctx.Output<Tensor>("ParamOut");
+
+    param_out->mutable_data<T>(ctx.GetPlace());
+
+    auto grad = ctx.Input<Tensor>("Grad");
```
Comment (lcy-seso) [inline paddle/operators/proximal_gd_op.h]: auto*

Actually using `auto` or `auto*` is not consistent in our codes. Some operators implemented very early use `auto` and some newly implemented operators use `auto*` when the return value is a pointer.

I prefer to use `auto*`, or at least we can keep consistent inside one operator.

File: paddle/operators/proximal_gd_op.h [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+
+    float l1 = ctx.Attr<float>("l1");
+    float l2 = ctx.Attr<float>("l2");
```
Comment (lcy-seso) [inline paddle/operators/proximal_gd_op.h]: I think here it is better to use

```cpp
auto l1 = static_cast<T>(ctx.Attr<float>("l1"));
auto l2 = static_cast<T>(ctx.Attr<float>("l2"));
```
or set l1 and l2 as a template parameter as in dropuout operator.

because `l1` and `l2` will be used in the calculation.  If we use double in future, the floats will be upconverted to doubles. But I wonder if we use half-precision or other lower precision, the final results will be as we expected?  so, maybe it is safer to cast l1 and l2 to T.

https://stackoverflow.com/questions/4239770/what-are-the-rules-governing-c-single-and-double-precision-mixed-calculations

File: paddle/operators/proximal_gd_op.cc (hunk 1 above)
Comment (kavyasrinet) [reply paddle/operators/proximal_gd_op.cc]: Sure, will do.

File: paddle/operators/proximal_gd_op.h (hunk 3 above)
Comment (kavyasrinet) [reply paddle/operators/proximal_gd_op.h]: I see. I used `auto` since the variable is being used in two places:
1. `auto g = EigenVector<T>::Flatten(*grad);`
2.  `Eigen::DSizes<int, 1> grad_dsize(grad->numel());`
And I could reuse the computation.
